code/datums/components/clothing_dirt.dm

```diff
@@ -0,0 +1,90 @@
+/// This component applies tint to clothing when its exposed to pepperspray, used in /obj/item/clothing/mask/gas.
+
+/datum/component/clothing_dirt
+	/// The ITEM_SLOT_* slot the item is equipped on, if it is.
+	var/equipped_slot
+
+	/// Mob wearing the clothing
+	var/mob/living/carbon/wearer
+
+	/// Amount of dirt stacks on the clothing
+	var/dirtiness = 0
+
+	/// Clothing we're applying tint on
+	var/obj/item/clothing/clothing
+
+/datum/component/clothing_dirt/Initialize()
+	if(!isclothing(parent))
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/clothing_dirt/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, PROC_REF(on_equip))
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, PROC_REF(on_drop))
+	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, PROC_REF(on_clean))
+	RegisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENTS, PROC_REF(on_expose))
+	clothing = parent
+
+/datum/component/clothing_dirt/UnregisterFromParent()
+	clothing.tint -= dirtiness
+	clothing = null
+	if (!isnull(wearer))
+		wearer.update_tint()
+		UnregisterSignal(wearer, COMSIG_ATOM_EXPOSE_REAGENTS)
+		wearer = null
+	else
+		UnregisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENTS)
+	UnregisterSignal(parent, list(
+		COMSIG_ATOM_EXAMINE,
+		COMSIG_ITEM_EQUIPPED,
+		COMSIG_MOB_UNEQUIPPED_ITEM,
+		COMSIG_COMPONENT_CLEAN_ACT,
+	))
+	return ..()
+
+/datum/component/clothing_dirt/proc/on_equip(datum/source, mob/user, slot)
+	SIGNAL_HANDLER
+	if (!(slot & (ITEM_SLOT_MASK | ITEM_SLOT_HEAD)))
+		return
```

Comment:
We could make this a bit more reusable if we just looked at the parent's slot flags
```dm
	var/obj/item/clothing/clothing = parent
	if (!(slot & clothing.slot_flags))
		return
```
Or you can let people set which slots it is active in via Initialize, we do this pattern often